src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -70,19 +81,133 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig)(implicit v
       try {
         executeGoogleRequest(request)
       } catch {
-        case e:GoogleJsonResponseException =>
-          if(e.getStatusCode!=404)
+        case e: GoogleJsonResponseException =>
+          if (e.getStatusCode != 404)
             throw CallToGoogleApiFailedException(googleProject, clusterName, e.getStatusCode, e.getDetails.getMessage)
       }
     }
   }
 
+  override def getClusterStatus(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[ClusterStatus] = {
+    getCluster(googleProject, clusterName).map { cluster =>
+      ClusterStatus.withNameIgnoreCase(cluster.getStatus.getState)
+    }
+  }
+
+  override def getClusterMasterInstanceIp(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[Option[String]] = {
+    val ipOpt = for {
+      cluster <- OptionT.liftF[Future, Cluster] { getCluster(googleProject, clusterName) }
+      masterInstanceName <- OptionT.fromOption { getMasterInstanceName(cluster) }
+      masterInstanceZone <- OptionT.fromOption { getZone(cluster) }
+      masterInstance <- OptionT.liftF[Future, Instance] { getInstance(googleProject, masterInstanceZone, masterInstanceName) }
+      masterInstanceIp <- OptionT.fromOption { getInstanceIP(masterInstance) }
+    } yield masterInstanceIp
+
+    ipOpt.value
+  }
+
+  override def getClusterErrorDetails(operationName: String)(implicit executionContext: ExecutionContext): Future[Option[ClusterErrorDetails]] = {
+    val errorOpt = for {
+      operation <- OptionT.liftF[Future, Operation] { getOperation(operationName) }
+      done <- OptionT.pure { operation.getDone } if done == true
+      error <- OptionT.pure { operation.getError }
+      code <- OptionT.pure { error.getCode }
+    } yield ClusterErrorDetails(code, Option(error.getMessage))
+
+    errorOpt.value
+  }
+
+  /**
+    * Gets a dataproc Cluster from the API.
+    */
+  private def getCluster(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[Cluster] = {
+    val request = dataproc.projects().regions().clusters().get(googleProject, dataprocConfig.dataprocDefaultRegion, clusterName)
+    Future {
+      executeGoogleRequest(request)
+    } recover {
+      case e: GoogleJsonResponseException =>
+        throw CallToGoogleApiFailedException(googleProject, clusterName, e.getStatusCode, e.getDetails.getMessage)
+    }
+  }
+
+  /**
+    * Gets a compute Instance from the API.
+    */
+  private def getInstance(googleProject: GoogleProject, zone: String, name: String)(implicit executionContext: ExecutionContext): Future[Instance] = {
+    val request = compute.instances().get(googleProject, zone, name)
+    Future {
+      executeGoogleRequest(request)
+    } recover {
+      case e: GoogleJsonResponseException =>
+        throw CallToGoogleApiFailedException(googleProject, name, e.getStatusCode, e.getDetails.getMessage)
+    }
```

Comment:
let's factor this pattern out, since we're using it all over the place

Reply:
Sounds good